zebra-chain/src/sapling/tests/prop.rs

```diff
@@ -157,12 +180,71 @@ proptest! {
         prop_assert_eq![data1, data2];
     }
 
+    /// Check that ShieldedData<SharedAnchor> is equal when `first` is swapped
+    /// between a spend and an output
+    #[test]
+    fn shielded_data_shared_swap_first_eq(
+        shielded1 in any::<sapling::ShieldedData<SharedAnchor>>()
+    ) {
+        use Either::*;
+
+        zebra_test::init();
+
+        // we need at least one spend and one output to swap them
+        prop_assume!(shielded1.spends().count() > 0 && shielded1.outputs().count() > 0);
+
+        let mut shielded2 = shielded1.clone();
+        let mut spends: Vec<_> = shielded2.spends().cloned().collect();
+        let mut outputs: Vec<_> = shielded2.outputs().cloned().collect();
+        match shielded2.first {
+            Left(_spend) => {
+                shielded2.first = Right(outputs.remove(0));
+                shielded2.rest_outputs = outputs;
+                shielded2.rest_spends = spends;
+            }
+            Right(_output) => {
+                shielded2.first = Left(spends.remove(0));
+                shielded2.rest_spends = spends;
+                shielded2.rest_outputs = outputs;
+            }
+        }
+
+        prop_assert_eq![&shielded1, &shielded2];
+
+        // shielded data doesn't serialize by itself, so we have to stick it in
+        // a transaction
+        let tx1 = Transaction::V5 {
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            lock_time: LockTime::min_lock_time(),
+            expiry_height: block::Height(0),
+            sapling_shielded_data: Some(shielded1),
+            rest: Vec::new()
+        };
+        let tx2 = Transaction::V5 {
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            lock_time: LockTime::min_lock_time(),
+            expiry_height: block::Height(0),
+            sapling_shielded_data: Some(shielded2),
+            rest: Vec::new()
+        };
```

Comment:
v5 shielded data should have a serialize implementation, because all its data is continuous.